lib/ruby_lsp/requests/support/rails_document_client.rb

```diff
@@ -0,0 +1,123 @@
+        )
+
+        class << self
+          extend T::Sig
+          sig do
+            params(name: String).returns(T::Array[String])
+          end
+          def generate_rails_document_urls(name)
+            docs = T.must(search_index)[name]
+
+            return [] unless docs
+
+            docs.map do |doc|
+              owner = doc[:owner]
+
+              link_name =
+                # class/module name
+                if owner == name
+                  name
+                else
+                  "#{owner}##{name}"
+                end
+
+              "[Rails Document: `#{link_name}`](#{doc[:url]})"
+            end
+          end
+
+          sig { returns(T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]])) }
+          private def search_index
+            @rails_documents ||= T.let(
+              build_search_index,
+              T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]]),
+            )
+          end
+
+          sig { returns(T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]])) }
+          private def build_search_index
+            return unless RAILTIES_VERSION
+
+            $stderr.puts "Fetching Rails Documents..."
+            # If the version's doc is not found, e.g. Rails main, it'll be redirected
+            # In this case, we just fetch the latest doc
+            response = if Gem::Version.new(RAILTIES_VERSION).prerelease?
+              Net::HTTP.get_response(URI("#{RAILS_DOC_HOST}/js/search_index.js"))
+            else
+              Net::HTTP.get_response(URI("#{RAILS_DOC_HOST}/v#{RAILTIES_VERSION}/js/search_index.js"))
+            end
+
+            if response.code == "200"
+              process_search_index(response.body)
+            else
+              $stderr.puts("Response failed: #{response.inspect}")
+              nil
+            end
+          rescue StandardError => e
+            $stderr.puts("Exception occurred when fetching Rails document index: #{e.inspect}")
+          end
+
+          sig { params(js: String).returns(T::Hash[String, T::Array[T::Hash[Symbol, String]]]) }
+          private def process_search_index(js)
+            table = {}
+            raw_data = js.sub("var search_data = ", "")
+            info = JSON.parse(raw_data).dig("index", "info")
+
+            # An entry looks like this:
+            #
+            # ["belongs_to",                                                              # method or module/class
+            #  "ActiveRecord::Associations::ClassMethods",                                # method owner
+            #  "classes/ActiveRecord/Associations/ClassMethods.html#method-i-belongs_to", # path to the document
+            #  "(name, scope = nil, **options)",                                          # method's parameters
+            #  "<p>Specifies a one-to-one association with another class..."]             # document preview
+            #
+            info.each do |ary|
```

Comment:
Oh, one small thing that's bugging me about this: I think we can make this code much more readable by expanding the array values here in the block params:
```suggestion
            info.each do |method_or_class, method_owner, doc_path, _, doc_preview|
```
which will allow us to get rid of the adhoc assignments with indices below inside the block.

Reply:
Great catch 👍